File: backend/app/schemas/rapports.py
from datetime import date
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel

from app.schemas.comptabilite import RapportFinancierResponse
from app.schemas.eleve import EffectifStatsResponse


class SerieGraphique(BaseModel):
    labels: list[str]
    values: list[float]


class DashboardKPIResponse(BaseModel):
    annee_libelle: str
    total_eleves: int
    total_classes: int
    total_personnel: int
    recettes_mois: Decimal
    total_impayes: Decimal
    taux_presence_mois: float | None
    nombre_impayes: int


class RapportEffectifsResponse(BaseModel):
    annee_libelle: str
    stats: EffectifStatsResponse
    par_classe: list[dict]


class RapportPedagogiqueClasseItem(BaseModel):
    classe_id: UUID
    classe_nom: str
    effectif: int
    moyenne_classe: Decimal | None
    taux_reussite: Decimal | None
    meilleur_eleve: str | None


class RapportPedagogiqueResponse(BaseModel):
    periode_id: UUID
    periode_libelle: str
    classes: list[RapportPedagogiqueClasseItem]


class RapportPresenceClasseItem(BaseModel):
    classe_id: UUID
    classe_nom: str
    effectif: int
    jours_absents: int
    jours_retards: int
    taux_presence: float | None


class RapportPresenceResponse(BaseModel):
    date_debut: date
    date_fin: date
    total_jours_suivis: int
    jours_absents_total: int
    jours_retards_total: int
    taux_presence_global: float | None
    par_classe: list[RapportPresenceClasseItem]


class StatistiquesAnnuellesResponse(BaseModel):
    etablissement: str
    annee_libelle: str
    total_eleves: int
    total_garcons: int
    total_filles: int
    total_classes: int
    total_recettes: Decimal
    total_depenses: Decimal
    solde_financier: Decimal
    moyenne_generale_etablissement: Decimal | None
    taux_reussite_global: Decimal | None
    taux_presence_annuel: float | None
    nombre_impayes: int


class GraphiquesResponse(BaseModel):
    effectifs_par_niveau: SerieGraphique
    recettes_par_mois: SerieGraphique
    repartition_sexe: SerieGraphique
    depenses_par_categorie: SerieGraphique


class RapportFinancierWrapper(BaseModel):
    rapport: RapportFinancierResponse

File: frontend/src/types/rapports.ts
export type DashboardKPI = {
  annee_libelle: string;
  total_eleves: number;
  total_classes: number;
  total_personnel: number;
  recettes_mois: string;
  total_impayes: string;
  taux_presence_mois: number | null;
  nombre_impayes: number;
};

export type SerieGraphique = {
  labels: string[];
  values: number[];
};

export type Graphiques = {
  effectifs_par_niveau: SerieGraphique;
  recettes_par_mois: SerieGraphique;
  repartition_sexe: SerieGraphique;
  depenses_par_categorie: SerieGraphique;
};

export type RapportEffectifs = {
  annee_libelle: string;
  stats: {
    total_eleves: number;
    total_garcons: number;
    total_filles: number;
    par_niveau: { niveau_code: string; niveau_libelle: string; total: number; garcons: number; filles: number }[];
    sans_classe: number;
  };
  par_classe: {
    id: string;
    nom: string;
    niveau_libelle: string;
    effectif: number;
    capacite_max: number;
    places_restantes: number;
  }[];
};

export type RapportPedagogique = {
  periode_id: string;
  periode_libelle: string;
  classes: {
    classe_id: string;
    classe_nom: string;
    effectif: number;
    moyenne_classe: string | null;
    taux_reussite: string | null;
    meilleur_eleve: string | null;
  }[];
};

export type RapportPresence = {
  date_debut: string;
  date_fin: string;
  total_jours_suivis: number;
  jours_absents_total: number;
  jours_retards_total: number;
  taux_presence_global: number | null;
  par_classe: {
    classe_id: string;
    classe_nom: string;
    effectif: number;
    jours_absents: number;
    jours_retards: number;
    taux_presence: number | null;
  }[];
};

export type StatistiquesAnnuelles = {
  etablissement: string;
  annee_libelle: string;
  total_eleves: number;
  total_garcons: number;
  total_filles: number;
  total_classes: number;
  total_recettes: string;
  total_depenses: string;
  solde_financier: string;
  moyenne_generale_etablissement: string | null;
  taux_reussite_global: string | null;
  taux_presence_annuel: number | null;
  nombre_impayes: number;
};

export type RapportFinancier = {
  periode_debut: string;
  periode_fin: string;
  total_recettes: string;
  total_depenses: string;
  solde: string;
  recettes_par_mode: Record<string, string>;
  depenses_par_categorie: Record<string, string>;
};
